src/main/java/co/cask/hydrator/KafkaBatchSource.java

```
@@ -0,0 +1,579 @@
+      }
+
+      for (TopicAndPartition topicAndPartition : topicAndPartitions) {
+        long latestOffset = latestOffsetResponse.offsets(topicAndPartition.topic(), topicAndPartition.partition())[0];
+        Long start;
+        byte[] tableStart = table.read(topicAndPartition.toString());
+        if (tableStart != null) {
+          start = Bytes.toLong(tableStart);
+        } else {
+          start = offsets.containsKey(topicAndPartition) ? offsets.get(topicAndPartition) - 1 : null;
+        }
+
+
+        long earliestOffset = start == null || start == -2
+          ? earliestOffsetResponse.offsets(topicAndPartition.topic(), topicAndPartition.partition())[0] : start;
+        if (earliestOffset == -1) {
+          earliestOffset = latestOffset;
+        }
+        LOG.info("earliest offset is {}", earliestOffset);
+        LOG.info("latest offset is {}", latestOffset);
+        //TODO: factor out kafka specific request functionality
+        KafkaRequest KafkaRequest =
+          new KafkaRequest(topicAndPartition.topic(), Integer.toString(leader.getLeaderId()),
+                           topicAndPartition.partition(), leader.getUri());
+        KafkaRequest.setLatestOffset(latestOffset);
+        KafkaRequest.setEarliestOffset(earliestOffset);
+        finalRequests.add(KafkaRequest);
+      }
+    }
+    return finalRequests;
+  }
+
+  private OffsetResponse getLatestOffsetResponse(SimpleConsumer consumer,
+                                                   Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetInfo) {
+    try {
+      OffsetResponse offsetResponse =
+        consumer.getOffsetsBefore(new kafka.javaapi.OffsetRequest(offsetInfo, kafka.api.OffsetRequest.CurrentVersion(),
+                                                                  "client"));
+      if (offsetResponse.hasError()) {
+        throw new RuntimeException("offsetReponse has error.");
+      }
+      return offsetResponse;
+    } catch (Exception e) {
+      //
```

ignoring an exception is almost never the right thing to do.  When can this happen? Shouldn't it cause the job to fail?